### PR TITLE
Droptest ino

### PR DESCRIPTION
**Pull request by PCModeActivate**
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7)

### CODE REVIEW

**Review by wendi-yu — COMMENTED** [review]
Few comments from me, mostly formatting and stuff but looks good!

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-:-MXIr2CJ5iwUu8IGP_vH:byvalh2)* status: 0 of 1 files reviewed, 8 unresolved discussions (waiting on @jacobdeery, @kylemtam, @PCModeActivate, and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 17 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInwLF9B-hhmF17fp7:-MXInwLF9B-hhmF17fp8:b-69uui4) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L17)):*
> ```cpp
> {
>   //start serial
>   Serial.begin(115200);
> ```

Is this just an arbitrary value? It looks a little magic-number-like, could you pull it out into a named variable or comment why it's 115200?
___
*[test_SD_card/test_SD_card.ino, line 18 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInZ-l7Wh0Mvu8_412:-MXInZ-l7Wh0Mvu8_413:blpcah5) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L18)):*
> ```cpp
>   //start serial
>   Serial.begin(115200);
>   
> ```

You've got some extraneous whitespace floating around here (and through the rest of the file), maybe run this through a code formatter? Jacob recently set up Astyle on some repo or other, you could try using that.
___
*[test_SD_card/test_SD_card.ino, line 27 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIqR8_1Cq4DOX4C09e:-MXIqR8_1Cq4DOX4C09f:b-dsb5rr) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L27)):*
> ```cpp
>   //setting CS pin to output
>   pinMode(10, OUTPUT);
>   pinMode(3, INPUT);
> ```

this one is being set as `INPUT` in the output section?
___
*[test_SD_card/test_SD_card.ino, line 37 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIqq2M3J3moTdQQEGF:-MXIqq2M3J3moTdQQEGG:b-bgzoq6) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L37)):*
> ```cpp
>   voltage = SD.open(file3, FILE_WRITE);
>  
>   // if the file opened okay, write to it:
> ```

Do we do anything if it didn't open ok lol, like stop the test? Just wondering
___
*[test_SD_card/test_SD_card.ino, line 58 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIov237hirxWbDTpuW:-MXIov237hirxWbDTpuX:bn4q028) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L58)):*
> ```cpp
>   OCR1A = 6250;            // compare match register 16MHz/256 Prescalar/1MHz
> 
>   TCCR1B |= (1 << WGM12);   // CTC mode
> ```

Nitpick but it's set to 0 right now, I don't think you need the `|`?
___
*[test_SD_card/test_SD_card.ino, line 78 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIpjT49o4saFzavO6E:-MXIpjT49o4saFzavO6F:bnivgag) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L78)):*
> ```cpp
> void loop()
> {
>   float diode_val = analogRead(A1)* 5.0/1023.0;
> ```

Maybe pull the `5.0/1023.0` out into a named constant? We use it a couple of times 
___
*[test_SD_card/test_SD_card.ino, line 81 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIpxqZEOfn4ifSWM7u:-MXIpxqZEOfn4ifSWM7v:b-5zfsyb) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L81)):*
> ```cpp
>   float sipm_val = analogRead(A0)* 5.0/1023.0;
>   float voltage_val = analogRead(A2)* 5.0/1023.0;
>   //Serial.print(diode_val, 6);
> ```

Probably best to just remove code rather than commenting it out, or stick it under an `if(debug_mode)` conditional (and just have `debug_mode` set at the top of the file)
___
*[test_SD_card/test_SD_card.ino, line 88 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIq9uU07zBIg-Gmifa:-MXIq9uU07zBIg-Gmifb:b-8l0k57) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L88)):*
> ```cpp
>   //Serial.println();
> 
>   diodeOutput.print(millis() );//+ i*3600);
> ```

same with the comment here

File: test_SD_card/test_SD_card.ino [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  interrupts();             // enable all interrupts
+  
+}
+
+ISR(TIMER1_COMPA_vect)
+{ 
+  //Interrupt Closes and reopens the files
+  savefile = true;
+}
```
Comment (zx100x100) [inline test_SD_card/test_SD_card.ino:74]: How often does this interrupt fire?

**Review by zx100x100 — COMMENTED** [review]
Looks good to me, but I'd wait for someone better with good formatting to take a look.

**Review by kylemtam — COMMENTED** [review]
I left a few questions (I apologize if the answers to the questions are quite apparent, I just don't follow the software-side of things too in-depth) but otherwise looks good!

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-:-MXdRRhA5s4r7H2DUYGc:brj9bb8)* status: 0 of 1 files reviewed, 9 unresolved discussions (waiting on @jacobdeery, @PCModeActivate, @wendi-yu, and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 37 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIqq2M3J3moTdQQEGF:-MXdP9u77wObnRHfLnwf:b-7tyvy5) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L37)):*
<details><summary><i>Previously, wendi-yu (Wendi Yu) wrote…</i></summary><blockquote>

Do we do anything if it didn't open ok lol, like stop the test? Just wondering
</blockquote></details>

Also just wondering but if file2 and file3 don't open, should we be stopping the test (or does it not matter if the files aren't there)?
___
*[test_SD_card/test_SD_card.ino, line 74 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MWR0VeF3nwwHegOBlQC-r5-74:-MXdQtGZCjJS6eBTfnUa:bz0r94w) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L74)):*
<details><summary><i>Previously, zx100x100 (Zachariah Mears) wrote…</i></summary><blockquote>

How often does this interrupt fire?
</blockquote></details>

Also I'm sure everybody else who reviewed this knows and it's quite apparent, but could I ask what triggers the interrupt? And also what is the purpose of closing and reopening the file after the interrupt fires?

**Review by PCModeActivate — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jacobdeery, @kylemtam, @wendi-yu, and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 17 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInwLF9B-hhmF17fp7:-MXTDdZhBsCusKDZZkVH:b1i5rtx) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L17)):*
<details><summary><i>Previously, wendi-yu (Wendi Yu) wrote…</i></summary><blockquote>

Is this just an arbitrary value? It looks a little magic-number-like, could you pull it out into a named variable or comment why it's 115200?
</blockquote></details>

The number was here when I was handed the code from Lana, I'm not sure why it is that value either.
___
*[test_SD_card/test_SD_card.ino, line 18 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInZ-l7Wh0Mvu8_412:-M_aRmO3CO1_q6f4ss5I:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L18)):*
<details><summary><i>Previously, wendi-yu (Wendi Yu) wrote…</i></summary><blockquote>

You've got some extraneous whitespace floating around here (and through the rest of the file), maybe run this through a code formatter? Jacob recently set up Astyle on some repo or other, you could try using that.
</blockquote></details>

done
___
*[test_SD_card/test_SD_card.ino, line 37 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXIqq2M3J3moTdQQEGF:-M_aR5HJ3NyPrvehciy6:b-vm9yhj) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L37)):*
<details><summary><i>Previously, kylemtam (Kyle Tam) wrote…</i></summary><blockquote>

Also just wondering but if file2 and file3 don't open, should we be stopping the test (or does it not matter if the files aren't there)?
</blockquote></details>

The desired effect when the file doesnt open is actually to flash the LED and print SD not okay
but since the code for that hasn't been tested I shouldnt be putting it here; instead, I'll put it in the next PR when it is added :)
___
*[test_SD_card/test_SD_card.ino, line 74 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MWR0VeF3nwwHegOBlQC-r5-74:-M_aMPbVCX0M7-qqYT0W:bsgcs9n) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L74)):*
<details><summary><i>Previously, kylemtam (Kyle Tam) wrote…</i></summary><blockquote>

Also I'm sure everybody else who reviewed this knows and it's quite apparent, but could I ask what triggers the interrupt? And also what is the purpose of closing and reopening the file after the interrupt fires?
</blockquote></details>

The interrupt fires in regular timer intervals (i.e. whenever the timer reaches the max timer value)
The current timer frequency is 10Hz
The purpose of closing and reopening the file is so that the data recorded in the output stream is saved into the file by closing the file (otherwise the data is lost)

The value of OCR1A is determined by CPU frequency (16MHz), Prescalar value (1/256) and desired frequency
In other words, OCR1A is the maximum timer value (it needs to be less than 65536 (16bit timer) to prevent overflow)
By setting OCR1A = 6250, we get desired frequency of 10Hz

**Review by wendi-yu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @jacobdeery, @PCModeActivate, and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 17 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInwLF9B-hhmF17fp7:-M_ihM6J4B9dajCSmXxo:b-vk4qyl) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L17)):*
<details><summary><i>Previously, PCModeActivate (Robert Cai) wrote…</i></summary><blockquote>

The number was here when I was handed the code from Lana, I'm not sure why it is that value either.
</blockquote></details>

Mind checking with Lana then? I'd ask but I have even less context than you lol

**Review by PCModeActivate — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @jacobdeery, @wendi-yu, and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 17 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MXInwLF9B-hhmF17fp7:-M_qHOh_9oXAi2WSb-Bz:b-kyfje2) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L17)):*
<details><summary><i>Previously, wendi-yu (Wendi Yu) wrote…</i></summary><blockquote>

Mind checking with Lana then? I'd ask but I have even less context than you lol
</blockquote></details>

Oh, I did and it turns out it was the Baud Rate, so I left my comment there

**Review by PCModeActivate — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jacobdeery and @zx100x100)

___
*[test_SD_card/test_SD_card.ino, line 74 at r5](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-MWR0VeF3nwwHegOBlQC-r5-74:-M_uyyID87jgfjLJ7g76:b-896fix) ([raw file](https://github.com/waterloo-rocketry/payload-2020/blob/ce08de54a66f7ff1800c14aba59262b43c156fc2/test_SD_card/test_SD_card.ino#L74)):*
<details><summary><i>Previously, PCModeActivate (Robert Cai) wrote…</i></summary><blockquote>

The interrupt fires in regular timer intervals (i.e. whenever the timer reaches the max timer value)
The current timer frequency is 10Hz
The purpose of closing and reopening the file is so that the data recorded in the output stream is saved into the file by closing the file (otherwise the data is lost)

The value of OCR1A is determined by CPU frequency (16MHz), Prescalar value (1/256) and desired frequency
In other words, OCR1A is the maximum timer value (it needs to be less than 65536 (16bit timer) to prevent overflow)
By setting OCR1A = 6250, we get desired frequency of 10Hz
</blockquote></details>

Done.

**Review by wendi-yu — APPROVED** [review]
Yep my comments look good

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/payload-2020/7#-:-MaWFw8SDgvZpjEvBnzU:b8qirvp)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jacobdeery and @zx100x100)
